types/fuzzaldrin/index.d.ts

```diff
@@ -2,7 +2,9 @@
 // Project: https://github.com/atom/fuzzaldrin
 // Definitions by: Mohamed Hegazy <https://github.com/mhegazy>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.1
 
-export function filter<T>(candidates: T[], query: string, options?: { key?: string, maxResults?: number }): T[];
+export function filter(candidates: string[], query: string, options?: {maxResults?: number}): string[];
+export function filter<T, K extends keyof T>(candidates: T[], query: string & T[K], options: {key: K, maxResults?: number}): T[];
```

Comment:
The intention with the `string & T[K]` part is to check that `T[K]` is a string (or subtype of string). Otherwise it should resolve to `never` and the call to filter() will fail to typecheck.

Reply:
`keyof T` is always a subtype of `string`, not sure what this is doing here really, do you have an example use case that needs this?

Reply:
`keyof T` is always a string, `T[keyof T]` is not.

The use case is to give an error for this invalid code:

```typescript
import { filter } from 'fuzzaldrin';
let objects: Array<{name: string, speed: number}> = [];

filter(objects, 'foo', {key: 'speed'});
```

The mistake is providing the key `'speed'` which points to a number, but the best way I could find to make it fail was to constrain on the `query` parameter.

Reply:
so is `filter(objects, 2, {key: 'speed'});` legal or not?

Reply:
nevermind, tried it, it is not legal. now i understand. thanks.